htm/src/nupic/types/Types.hpp

```diff
@@ -0,0 +1,310 @@
+#pragma warning(                                                               \
+    disable : 4305) // truncation from 'double' to 'nta::Real', possible loss of
+                    // data (LOTS of various type combinations)
+#pragma warning(once : 4838) // narrowing conversions
+#endif
+
+#define UNUSED(x) (void)(x)  // use this to avoid 'unused variable' compile errors.
+#define NTA_INTERNAL 1       // just means not part of API even if exposed as public.
+
+
+#if defined(NTA_OS_WINDOWS)
+#define NTA_EXPORT __declspec(dllexport)
+#define NTA_HIDDEN
+#else
+#define NTA_EXPORT __attribute__((visibility("default")))
+#define NTA_HIDDEN __attribute__((visibility("hidden")))
+#endif
+
+
+
+//----------------------------------------------------------------------
+
+namespace nupic {
+
+///////////////////////// Basic types ///////////////////////////////////////
+
+/**
+ * @name Basic types
+ *
+ * @{
+ */
+
+/**
+ * Represents a 8-bit byte.
+ */
+typedef char Byte;
+
+/**
+ * Represents a 16-bit signed integer.
+ */
+typedef short Int16;
+
+/**
+ * Represents a 16-bit unsigned integer.
+ */
+typedef unsigned short UInt16;
+
+/**
+ * Represents a 32-bit signed integer.
+ */
+typedef long Int32;
+
+/**
+ * Represents a 32-bit unsigned integer.
+ */
+typedef unsigned long UInt32;
+
+/**
+ * Represents a 64-bit signed integer.
+ */
+typedef long long Int64;
+
+/**
+ * Represents a 64-bit unsigned integer.
+ */
+typedef unsigned long long UInt64;
+
+/**
+ * Represents a 32-bit real number(a floating-point number).
+ */
+typedef float Real32;
+
+/**
+ * Represents a 64-bit real number(a floating-point number).
+ */
+typedef double Real64;
+
+/**
+ * Represents an opaque handle/pointer, same as `void *`
+ */
+typedef void* Handle;
+
+/**
+ * Represents an opaque pointer, same as `uintptr_t`
+ */
+typedef uintptr_t UIntPtr;
+
+/**
+ * Represents lengths of arrays, strings and so on.
+ */
+typedef size_t Size;
+
+
+
+/**
+ * @}
+ */
+
+/**
+ * @name Flexible types
+ *
+ * The following are flexible types depending on `NTA_DOUBLE_PRECISION` and
+ * `NTA_BIG_INTEGER`.
+ *
+ * @{
+ *
+ */
+
+/**
+ * Represents a real number(a floating-point number).
+ *
+ * Same as nupic::Real64 if `NTA_DOUBLE_PRECISION` is defined, nupic::Real32 otherwise.
+ */
+#ifdef NTA_DOUBLE_PRECISION
+	typedef Real64 Real;
+#else
+	typedef Real32 Real;
+#endif
+/**
+ * Represents a signed integer.
+ *
+ * Same as nupic::Int64 if `NTA_BIG_INTEGER` is defined, nupic::Int32 otherwise.
+ */
+#ifdef NTA_BIG_INTEGER
+	typedef Int64 Int;
+#else
+	typedef Int32 Int;
+#endif
+
+/**
+ * Represents a unsigned integer.
+ *
+ * Same as nupic::UInt64 if `NTA_BIG_INTEGER` is defined, nupic::UInt32
+ * otherwise.
+ */
+#ifdef NTA_BIG_INTEGER
+	typedef Uint64 UInt;
+#else
+	typedef UInt32 UInt;
+#endif
+
+
+
+///////////////////////// Basic type enums ///////////////////////////////////////
+
+/**
+ * Basic types enumeration
+ */
+typedef enum NTA_BasicType {
```

Comment:
OT: but are we eventually aiming to remove the `NTA_BasicType_*` ?

Reply:
> OT: but are we eventually aiming to remove the NTA_BasicType_* ?

No... NTA_BasicType is used extensively in NetworkAPI.
In fact I want to extend this by adding an NTA_BasicTypeSDR so I will be able to wrap the SDR class inside the Array class so we can pass an SDR between Regions using the Links.

I will setup another issue to discuss my thinking on this.